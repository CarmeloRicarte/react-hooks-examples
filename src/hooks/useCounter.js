import { useState } from "react"

export const useCounter = (initialValue = 10) => {
    const [counter, setCounter] = useState(initialValue);

    /* It takes a value as an argument, and adds the value to the counter. */
    const increment = (value = 1) => setCounter(counter + value);

    /**
     * It takes a value as an argument, and if the counter is greater than 0, it subtracts the value
     * from the counter
     * @param [value=1] - The value to decrement the counter by.
     * @returns the value of the counter.
     */
    const decrement = (value = 1) => {
        if (counter === 0) return;
        setCounter(counter - value);
    }
    const resetCounter = () => setCounter(initialValue);

    return {
        counter,
        increment,
        decrement,
        resetCounter
    }

}
