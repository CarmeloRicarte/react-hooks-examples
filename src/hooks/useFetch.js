import { useEffect, useState } from "react"

/**
 * We're using the useEffect hook to make a fetch request to the url we defined earlier. If the request
 * is successful, we set the data in state to the data we received from the request. If the request
 * fails, we set the error in state to the error we received
 * @param url - The url we want to fetch data from
 * @returns {
 *         data: state.data,
 *         isLoading: state.isLoading,
 *         error: state.error
 *     }
 */
export const useFetch = (url) => {
    const [state, setState] = useState({
        data: null,
        isLoading: true,
        error: null
    })

    const getFetch = async () => {
        setState({
            ...state,
            isLoading: true
        })

        try {
            const response = await fetch(url)
            const data = await response.json();
            setState({ data, isLoading: false, error: null })

        } catch (err) {
            setState({
                ...state,
                isLoading: false,
                error: err,
            })
        }

    }

    useEffect(() => {
        getFetch();
    }, [url])

    return {
        data: state.data,
        isLoading: state.isLoading,
        error: state.error
    }
}
