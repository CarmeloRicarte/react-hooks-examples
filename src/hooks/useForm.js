import { useState } from "react";

export const useForm = (initialForm = {}) => {
  const [formState, setFormState] = useState(initialForm);

  /**
   * When the input changes, update the state of the form with the new value of the input.
   */
  const onInputChange = ({ target }) => {
    const { name, value } = target;
    setFormState({
      ...formState, // maintain state of other form fields
      [name]: value, // update state of field that is changed
    });
  };

  /**
   * It resets the form to its initial state.
   */
  const onResetForm = () => {
    setFormState(initialForm);
  };

  return {
    ...formState, // for return every form control of state
    formState,
    onInputChange,
    onResetForm,
  };
};
