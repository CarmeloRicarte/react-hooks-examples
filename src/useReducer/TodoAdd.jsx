import { useRef } from "react";
import PropTypes from "prop-types";
const TodoAdd = ({ onNewTodo }) => {
  const inputRef = useRef();
  /**
   * When the form is submitted, prevent the default action, then if the input has a value, create a
   * new todo object with the current time as the id, the input value as the description, and done set
   * to false, then call the onNewTodo function with the new todo object as the argument
   */
  const handleSubmit = (event) => {
    event.preventDefault();
    if (inputRef.current.value) {
      const todo = {
        id: new Date().getTime(),
        description: inputRef.current.value,
        done: false,
      };
      onNewTodo(todo);
      inputRef.current.value = ""; // clear previous value of the input
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        ref={inputRef}
        type="text"
        placeholder="What needs to be done?"
        className="form-control"
      />

      <button type="submit" className="btn btn-outline-primary mt-2">
        Add
      </button>
    </form>
  );
};

TodoAdd.propTypes = {
  onNewTodo: PropTypes.func.isRequired,
};

export default TodoAdd;
