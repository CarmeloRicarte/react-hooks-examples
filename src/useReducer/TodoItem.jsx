import PropTypes from "prop-types";

const TodoItem = ({ todo, onDeleteTodo, onToggleTodo }) => {
  return (
    <li className="list-group-item d-flex justify-content-between my-2">
      <span
        data-testid={`${todo.id}-item-description`}
        onClick={() => onToggleTodo(todo.id)}
        className={`align-self-center ${
          todo.done ? "text-decoration-line-through" : ""
        }`}
      >
        {todo.description}
      </span>
      <button
        type="button"
        className="btn btn-danger"
        onClick={() => onDeleteTodo(todo.id)}
      >
        Delete
      </button>
    </li>
  );
};

TodoItem.propTypes = {
  todo: PropTypes.shape({
    id: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    done: PropTypes.bool.isRequired,
  }),
  onDeleteTodo: PropTypes.func.isRequired,
  onToggleTodo: PropTypes.func.isRequired,
};

export default TodoItem;
