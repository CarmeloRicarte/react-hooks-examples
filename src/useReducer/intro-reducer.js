const initialState = [{
    id: 1,
    todo: 'Learn React',
    done: false
}]

const newTodo = {
    id: 2,
    todo: 'Learn React Query',
    done: false
}

const ADD_TODO = {
    type: 'ADD_TODO',
    payload: newTodo
}

const todoReducer = (state = initialState, action = {}) => {
    switch (action.type) {
        case 'ADD_TODO':
            return [...state, action.payload];
        case 'REMOVE_TODO':
            return state.filter(todo => todo.id !== action.id);
        case 'UPDATE_TODO':
            return state.map(todo => {
                if (todo.id === action.id) {
                    return {
                        ...todo,
                        done: action.done
                    }
                }
                return todo;
            });
        default: throw new Error('Unknown action');
    }
}

console.log(todoReducer(initialState, ADD_TODO));
