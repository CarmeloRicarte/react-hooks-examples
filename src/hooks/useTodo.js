import { useEffect, useReducer } from "react";
import { todoReducer } from "../useReducer/todoReducer";


const init = () =>
    localStorage.getItem("todos") ? JSON.parse(localStorage.getItem("todos")) : []

export const useTodo = () => {
    const [todos, dispatchTodoAction] = useReducer(
        todoReducer,
        [],
        init
    );

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos]);

    /**
     * When the user submits a new todo, dispatch an action to add the new todo to the list of todos.
     * @param newTodo - The new todo item that we want to add to the list.
     */
    const handleNewTodo = (newTodo) => {
        dispatchTodoAction({
            type: "ADD_TODO",
            payload: newTodo,
        });
    };

    /**
     * It takes an id as an argument, and then dispatches an action to the reducer with the type of
     * "REMOVE_TODO" and the payload of the id
     * @param id - The id of the todo item that we want to delete.
     */
    const handleDeleteTodo = (id) => {
        dispatchTodoAction({
            type: "REMOVE_TODO",
            payload: id,
        });
    };

    /**
     * It takes an id as an argument, and then dispatches an action to the reducer with the type
     * "MARK_AS_DONE" and the id as the payload
     * @param id - The id of the todo item that we want to mark as done.
     */
    const handleToggleTodo = (id) => {
        dispatchTodoAction({
            type: "MARK_AS_DONE",
            payload: id,
        });
    };

    return {
        todos,
        handleNewTodo,
        handleDeleteTodo,
        handleToggleTodo,
        todosCount: todos.length,
        pendingTodosCount: todos.filter((todo) => todo.done === false).length
    }
}

