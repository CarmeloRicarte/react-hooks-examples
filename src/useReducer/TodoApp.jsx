import { useTodo } from "../hooks/useTodo";
import TodoAdd from "./TodoAdd";
import TodoList from "./TodoList";

export const TodoApp = () => {
  // useTodo con todos, los handle, itialState, init
  const {
    todos,
    handleNewTodo,
    handleDeleteTodo,
    handleToggleTodo,
    todosCount,
    pendingTodosCount,
  } = useTodo();

  return (
    <>
      <h1>TODO App: {todosCount}</h1>
      <h2>Pendientes: {pendingTodosCount}</h2>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            onDeleteTodo={handleDeleteTodo}
            onToggleTodo={handleToggleTodo}
          ></TodoList>
        </div>

        <div className="col-5">
          <h4>Add TODO</h4>
          <hr />
          <TodoAdd onNewTodo={handleNewTodo} />
        </div>
      </div>
    </>
  );
};
