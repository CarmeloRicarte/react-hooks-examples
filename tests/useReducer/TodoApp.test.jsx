import { render, screen } from "@testing-library/react";
import { expect, vi } from "vitest";
import { useTodo } from "../../src/hooks/useTodo";
import { TodoApp } from "../../src/useReducer/TodoApp";

vi.mock("../../src/hooks/useTodo");
describe("Test of <TodoApp />", () => {
  useTodo.mockReturnValue({
    todos: [
      {
        id: 1,
        description: "Todo 1",
        done: false,
      },
      {
        id: 2,
        description: "Todo 2",
        done: true,
      },
    ],
    handleNewTodo: vi.fn(),
    handleDeleteTodo: vi.fn(),
    handleToggleTodo: vi.fn(),
    todosCount: 2,
    pendingTodosCount: 1,
  });
  test("should match with snapshot", () => {
    const { container } = render(<TodoApp />);
    expect(container).toMatchSnapshot();
  });

  test("should show component correctly", () => {
    render(<TodoApp />);
    expect(screen.getByText("Todo 1")).toBeTruthy();
    expect(screen.getByText("Todo 2")).toBeTruthy();
    expect(screen.getByRole("textbox")).toBeTruthy();
  });
});
