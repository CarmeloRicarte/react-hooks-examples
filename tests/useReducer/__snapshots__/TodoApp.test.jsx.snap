// Vitest Snapshot v1

exports[`Test of <TodoApp /> > should match with snapshot 1`] = `
<div>
  <h1>
    TODO App: 
    2
  </h1>
  <h2>
    Pendientes: 
    1
  </h2>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between my-2"
        >
          <span
            class="align-self-center "
            data-testid="1-item-description"
          >
            Todo 1
          </span>
          <button
            class="btn btn-danger"
            type="button"
          >
            Delete
          </button>
        </li>
        <li
          class="list-group-item d-flex justify-content-between my-2"
        >
          <span
            class="align-self-center text-decoration-line-through"
            data-testid="2-item-description"
          >
            Todo 2
          </span>
          <button
            class="btn btn-danger"
            type="button"
          >
            Delete
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Add TODO
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          placeholder="What needs to be done?"
          type="text"
        />
        <button
          class="btn btn-outline-primary mt-2"
          type="submit"
        >
          Add
        </button>
      </form>
    </div>
  </div>
</div>
`;
